libs/platform/ebpf_epoch.c

```diff
@@ -137,100 +191,71 @@ typedef struct _ebpf_epoch_allocation_header
-    _Inout_ ebpf_epoch_cpu_entry_t* cpu_entry);
+_Function_class_(KDEFERRED_ROUTINE) _IRQL_requires_(DISPATCH_LEVEL) static void _ebpf_epoch_messenger_worker(
+    _In_ KDPC* dpc, _In_opt_ void* cpu_entry, _In_opt_ void* message, _In_opt_ void* arg2);
 
-/**
- * @brief Get the CPU ID from the epoch state.
- *
- * @param[in] epoch_state The epoch state to get the CPU ID from.
- * @return The CPU ID.
- */
-uint32_t
-_ebpf_epoch_get_cpu_id_from_state(_In_ ebpf_epoch_state_t* epoch_state);
+_Function_class_(KDEFERRED_ROUTINE) _IRQL_requires_(DISPATCH_LEVEL) static void _ebpf_epoch_timer_worker(
+    _In_ KDPC* dpc, _In_opt_ void* cpu_entry, _In_opt_ void* message, _In_opt_ void* arg2);
 
-/**
- * @brief Remove all entries from the per-CPU free list that have an epoch that is before released_epoch.
- *
- * @param[in] cpu_id The per-CPU free list to search.
- * @param[in] released_epoch The epoch to release.
- */
-static void
-_ebpf_epoch_release_free_list(_Inout_ ebpf_epoch_cpu_entry_t* cpu_entry, int64_t released_epoch);
+_IRQL_requires_max_(APC_LEVEL) static void _ebpf_epoch_send_message_and_wait(
+    _In_ ebpf_epoch_cpu_message_t* message, uint32_t cpu_id);
 
-/**
- * @brief Determine the newest inactive epoch and return it.
- *
- * @param[out] release_epoch The newest inactive epoch.
- */
 static void
-_ebpf_epoch_get_release_epoch(_Out_ int64_t* released_epoch);
+_ebpf_epoch_insert_in_free_list(_In_ ebpf_epoch_allocation_header_t* header);
+
+static _IRQL_requires_(DISPATCH_LEVEL) void _ebpf_epoch_arm_timer_if_needed(ebpf_epoch_cpu_entry_t* cpu_entry);
 
-/**
- * @brief Routine executed on a timer to compute the newest inactive epoch.
- *
- * @param[in] context Unused.
- */
 static void
-_ebpf_flush_worker(_In_ const void* context);
+_ebpf_epoch_work_item_callback(void* context);
 
 /**
- * @brief Flush any stale entries from the per-CPU free list.
+ * @brief Raise the CPU's IRQL to DISPATCH_LEVEL if it is below DISPATCH_LEVEL.
+ * First check if the IRQL is below DISPATCH_LEVEL to avoid the overhead of
+ * calling KeRaiseIrqlToDpcLevel() if it is not needed.
  *
- * @param[in] dpc Unused.
- * @param[in] deferred_context Unused.
- * @param[in] parameter_1 Unused.
- * @param[in] parameter_2 Unused.
+ * @return The previous IRQL.
  */
-static KDEFERRED_ROUTINE _ebpf_epoch_stale_worker;
+_IRQL_requires_max_(DISPATCH_LEVEL) _IRQL_saves_ _IRQL_raises_(DISPATCH_LEVEL) static inline KIRQL
+    _ebpf_epoch_raise_to_dispatch_if_needed()
+{
+    KIRQL current_irql = KeGetCurrentIrql();
+    if (current_irql < DISPATCH_LEVEL) {
+        return KeRaiseIrqlToDpcLevel();
+    }
+    return current_irql;
+}
 
 /**
- * @brief Arm the flush timer if:
- *  Timer is not already armed.
- *  Timer is not disabled.
- *  Free list is not empty.
+ * @brief Lower the CPU's IRQL to the previous IRQL if it is below DISPATCH_LEVEL.
+ * First check if the IRQL is below DISPATCH_LEVEL to avoid the overhead of
+ * calling KeLowerIrql() if it is not needed.
+ *
```

Comment:
shouldnt it check if the current irql is < DISPATCH?

Reply:
Added correct annotation to make it clear this must be called at dispatch.